wled00/FX_fcn.cpp

```diff
@@ -466,6 +462,15 @@ CRGBPalette16 IRAM_ATTR &Segment::currentPalette(CRGBPalette16 &targetPalette, u
 void Segment::handleRandomPalette() {
   // just do a blend; if the palettes are identical it will just compare 48 bytes (same as _randomPalette == _newRandomPalette)
   // this will slowly blend _newRandomPalette into _randomPalette every 15ms or 8ms (depending on MIN_SHOW_DELAY)
+  // if palette transitions is enabled, blend it according to Transition Time (if longer than minimum given by service calls) 
+  if(strip.paletteFade)
+  {
+    if((millis()&0xFFFF) - _lastPaletteBlend < strip.getTransition()>>7) //assumes that 128 updates are needed to blend a palette, so shift by 7 (can be more, can be less)
```

Comment:
Actually my measurements were 255 blends for full palette blend if colors were on the opposite ends.
I do not like this even though it adds predictability. After all do you really need *exact* time for two palettes to entirely blend? IMO not and I'd rather have CPU cycles spent elsewhere.

Reply:
yes, 'worst case' is 255 but typical much lower but since palette blending can be anything, it is better to pick an average value. also makes it easy to use bitshifts and save some ram. 
your comment is unclear to me. this does not add predictability but adds the possibility to fade more slowly. not needed for disco style lighting but very useful for slow ambient lights.

Reply:
This code:
```c++
  if (doShow) {
    yield();
    Segment::handleRandomPalette();
    show();
  }
```
introduced in one of recent commits ensures that palettes will be entirely blended in exactly 255 frames (which is ~6s assuming 42 FPS). Or, using your assumption of not entirely divergent palettes, in 128 frames (~3s).
If you reduce FPS you'll get even longer blend times. Assuming ambient light doesn't need high FPS.

If you think this is too fast then perhaps we need to introduce some scaling mechanism? As most users (until they become experienced) will use default transition time which is 0.7s making your code almost useless and producing almost instantaneous palette change. I do not want that.

Reply:
you are wrong on that ;) 
the code delays it by skipping the  nblendPaletteTowardPalette() if transition time is set higher than function call intervals. if it set lower, it behaves exactly as before, so transitioning in ~128 frames. it will never speed it up.

Reply:
"*ensures that palettes will be entirely blended in exactly 255 frames*"

A call to `show()` determines 1 frame. So the above statement is as true as it can be as each call to `Segment::handleRandomPalette()` will only blend 1/255th of intended blend.

`nblendPaletteTowardPalette()` will internally only compare 48 bytes when nothing needs to be done using `memcmp()`. This is *always* performed.

There is the other question/desire we need to clarify. What I want is a smooth blend of two random palettes which may change at user's desired time (anything between 1s and 65535s) and you want this smooth blend to happen within the *transition* time as specified in UI. These two desires have nothing in common (except for the smoothness).

You are mentioning slow, ambient illumination, blend which would also require longer period between random palette change (you wouldn't want to change random palette ever second and have transition time of 6553 seconds). So these two would need to be correlated somehow in such case.

As for CPU cycles: `loadPalette()` is called each time `color_from_palette()` is called which may be in 1000s of times *per frame*. You added two divisions, bitwise AND and subtraction to it. `Segment::handleRandomPalette()` is called once per frame.

Reply:
you make good points. remember I requested to have a separate value in the LED config for palette transitions? you are describing exactly my thought process for that request. palette change time and transition time should not be correlated I agree, I would much rather have a separate palette transition value, it would also decouple FX transitions from color/palette transitions. 
as for CPU cycles: you are absolutely correct. but the code was really bad to begin with, I just made it a tad worse.
I see no reason why that if statement needed to be in the `loadPalette()` fuction. 
So I now moved the if statement to the `Segment::handleRandomPalette()` function, so only called once per frame. It works just fine that way. Still need to implement the other updates discussed. will push it as soon as it is done.

Reply:
Discussion is the best way to solve problems (though it can make some as well). 😄